keps/sig-api-machinery/20190228-priority-and-fairness.md

```diff
@@ -0,0 +1,343 @@
+CoDel is a technique for ejecting requests from a queue for the
+purpose of keeping latency low, and fq_codel applies the CoDel
+technique in each of many queues.  CoDel is explicitly designed to
+work in the context of TCP flows on the Internet.  This KEP should be
+similarly explicit about the context, particularly including what is
+the feedback given to clients and how do they react and what is the
+net effect of all the interacting pieces.  No such analysis has yet
+been done for any of the proposals.
+
+The CoDel technique is described as parameterless but has two magic
+numbers: an "initial interval" of 100 ms and a "target" of 5 ms.  The
+initial interval is set based on round trip times in the Internet, and
+the target is set based on a desired limit on the latency at each hop.
+What are the analogous numbers for our scenario?  We do not have large
+numbers of hops; typically at most two (client to main apiserver and
+then main apiserver to aggregated apiserver).  What is analogous to
+network round trip time?  We have a latency goal of 1 second, and a
+request service time limit of 1 minute.  If we take 1 second as the
+initial interval then note that the maximum service time is much
+larger than the initial interval; by contrast, in networking, the
+maximum service time (i.e., packet length / link speed) is much
+smaller than the initial interval.  Even if we take 1 minute as our
+initial interval, we still do not have the sort of relationship that
+obtains in networking.  Note that in order to get good statistics on a
+queue --- which is needed by the CoDel technique --- there have to be
+many requests served during an interval.  Because of this mismatch,
+and because equivalence of context has not been established, we are
+not agreed that the CoDel technique can be used.
+
+Despite the open issues above, we seem to be circling around the
+following outline.
+
+- When a request arrives at the handler, the request is categorized
+  somehow.  Some proposals call for the request to be rejected upon
+  arrival based on that categorization and some local state.  Unless
+  rejected, the request is put into a FIFO queue.  That is one of many
+  queues.  The queues are associated with the categories somehow.
+  Some proposlas contemplate ejecting less desirable requests to make
+  room for the newly queued request, if and when queue space is tight.
+
+- A request might also be rejected at a later time, based on other
+  criteria.  For example, as in the CoDel technique --- which will
+  reject a request at the head of the queue if the latency is found to
+  be too big at certain times.
+
+- Based on some resource limit (i.e., QPS or concurrency), requests
+  are dispatched from the queues to be served (i.e., continue down the
+  handler chain).
+
+- We assume that when the requst-timeout handler aborts a request it
+  is effective --- we assume the request stops consuming CPU and
+  memory at that point.
```

Comment:
Actually not currently a valid assumption, something we need to fix.

Reply:
yes, that is part of what I meant by "we assume".  I will revise to be more clear.